File: go/internal/slhdsa/hypertree.go
package slhdsa

// Hypertree implementation per FIPS 205.

// HTSign generates a hypertree signature.
// FIPS 205 Algorithm 16: ht_sign(M, SK.seed, PK.seed, idx_tree, idx_leaf).
func HTSign(hs HashSuite, msg, skSeed, pkSeed []byte, idxTree uint64, idxLeaf int, p *Params) []byte {
	var adrs Address

	// Sign at layer 0
	adrs.SetLayerAddress(0)
	adrs.SetTreeAddress(idxTree)
	sigTmp := XMSSSign(hs, msg, skSeed, pkSeed, idxLeaf, &adrs, p)
	sigHT := make([]byte, 0, p.D*(p.Len+p.HP)*p.N)
	sigHT = append(sigHT, sigTmp...)

	root := XMSSPKFromSig(hs, idxLeaf, sigTmp, msg, pkSeed, &adrs, p)

	for j := 1; j < p.D; j++ {
		idxLeaf = int(idxTree & uint64((1<<uint(p.HP))-1))
		idxTree = idxTree >> uint(p.HP)

		adrs.SetLayerAddress(uint32(j))
		adrs.SetTreeAddress(idxTree)

		sigTmp = XMSSSign(hs, root, skSeed, pkSeed, idxLeaf, &adrs, p)
		sigHT = append(sigHT, sigTmp...)

		if j < p.D-1 {
			root = XMSSPKFromSig(hs, idxLeaf, sigTmp, root, pkSeed, &adrs, p)
		}
	}

	return sigHT
}

// HTVerify verifies a hypertree signature.
// FIPS 205 Algorithm 17: ht_verify(M, SIG_HT, PK.seed, idx_tree, idx_leaf, PK.root).
func HTVerify(hs HashSuite, msg, sigHT, pkSeed []byte, idxTree uint64, idxLeaf int, pkRoot []byte, p *Params) bool {
	var adrs Address
	n := p.N
	xmssLen := (p.Len + p.HP) * n

	adrs.SetLayerAddress(0)
	adrs.SetTreeAddress(idxTree)

	sigXMSS := sigHT[:xmssLen]
	node := XMSSPKFromSig(hs, idxLeaf, sigXMSS, msg, pkSeed, &adrs, p)

	for j := 1; j < p.D; j++ {
		idxLeaf = int(idxTree & uint64((1<<uint(p.HP))-1))
		idxTree = idxTree >> uint(p.HP)

		adrs.SetLayerAddress(uint32(j))
		adrs.SetTreeAddress(idxTree)

		sigXMSS = sigHT[j*xmssLen : (j+1)*xmssLen]
		node = XMSSPKFromSig(hs, idxLeaf, sigXMSS, node, pkSeed, &adrs, p)
	}

	// Compare with root
	if len(node) != len(pkRoot) {
		return false
	}
	for i := range node {
		if node[i] != pkRoot[i] {
			return false
		}
	}
	return true
}

File: swift/Sources/PQCStandards/SlhDsa/Hypertree.swift
// SLH-DSA Hypertree

public enum Hypertree {
    /// Hypertree Sign
    public static func sign(params: SlhDsaParams, msg: [UInt8], skSeed: [UInt8],
                           pkSeed: [UInt8], idxTree: UInt64, idxLeaf: UInt32) -> [UInt8] {
        var adrs = SlhAddress()

        // Sign at layer 0
        adrs.setLayerAddress(0)
        adrs.setTreeAddressBytes(idxTree)
        var sig = Xmss.sign(params: params, msg: msg, skSeed: skSeed,
                           pkSeed: pkSeed, idx: idxLeaf, adrs: &adrs)

        var root = Xmss.rootFromSig(params: params, sig: sig, msg: msg,
                                     pkSeed: pkSeed, idx: idxLeaf, adrs: &adrs)

        // Sign at higher layers
        var tree = idxTree
        var leaf = UInt32(tree & UInt64((1 << params.hPrime) - 1))

        for j in 1..<params.d {
            tree >>= params.hPrime
            leaf = UInt32(tree & UInt64((1 << params.hPrime) - 1))

            adrs.setLayerAddress(UInt32(j))
            adrs.setTreeAddressBytes(tree >> params.hPrime)

            let layerSig = Xmss.sign(params: params, msg: root, skSeed: skSeed,
                                     pkSeed: pkSeed, idx: leaf, adrs: &adrs)
            sig.append(contentsOf: layerSig)

            if j < params.d - 1 {
                root = Xmss.rootFromSig(params: params, sig: layerSig, msg: root,
                                         pkSeed: pkSeed, idx: leaf, adrs: &adrs)
            }
        }
        return sig
    }

    /// Hypertree Verify
    public static func verify(params: SlhDsaParams, msg: [UInt8], sig: [UInt8],
                             pkSeed: [UInt8], pkRoot: [UInt8],
                             idxTree: UInt64, idxLeaf: UInt32) -> Bool {
        var adrs = SlhAddress()
        let sigBlockSize = (params.len + params.hPrime) * params.n

        // Verify at layer 0
        adrs.setLayerAddress(0)
        adrs.setTreeAddressBytes(idxTree)
        let sig0 = Array(sig[0..<sigBlockSize])
        var node = Xmss.rootFromSig(params: params, sig: sig0, msg: msg,
                                     pkSeed: pkSeed, idx: idxLeaf, adrs: &adrs)

        var tree = idxTree
        var sigOff = sigBlockSize

        for j in 1..<params.d {
            tree >>= params.hPrime
            let leaf = UInt32(tree & UInt64((1 << params.hPrime) - 1))

            adrs.setLayerAddress(UInt32(j))
            adrs.setTreeAddressBytes(tree >> params.hPrime)

            let layerSig = Array(sig[sigOff..<(sigOff + sigBlockSize)])
            sigOff += sigBlockSize

            node = Xmss.rootFromSig(params: params, sig: layerSig, msg: node,
                                     pkSeed: pkSeed, idx: leaf, adrs: &adrs)
        }

        return node == pkRoot
    }
}
